### DISCUSSION TITLE
Only Reload Relations When Necessary

### DISCUSSION THREAD

**Discussion by mac-s-g** [spatie/laravel-activitylog#964] [Q&A] [answered]
Most importantly: thanks for building, sharing, and maintaining such an awesome project!

I've recently been working on minimizing database traffic in my application.  While profiling, I found that the `LogsActivity` trait is responsible for a high volume of (what I would consider) redundant database queries.

Each time a model emits an event defined by `static::$recordEvents` (created, updated, & deleted by default), the `attributeValuesToBeLogged()` method is invoked which reloads model relations via `$this->fresh()`.  
Here's a [link to that line of code](https://github.com/spatie/laravel-activitylog/blob/1d8acdac3e1f3d6937f8d10fd005de71b0350bc4/src/Traits/LogsActivity.php#L262).

If I understand correctly, Invoking the `fresh()` method on a model instance will result in N database queries, where N is the number of relations currently loaded on the model instance.  

The `LogsActivity` trait calls the `fresh()` method regardless of whether relations are updated.  Depending on the application and scale, that behavior can result in a very high volume of unnecessary database queries.  This has proven expensive for my application.

Of course, the `fresh()` call is useful in cases where relations are updated.  Here's a [link to the original discussion](https://github.com/spatie/laravel-activitylog/pull/158) on adding it.

My goal is mainly to identify this issue and start a discussion about possible optimizations.

Would it be feasible to prevent unnecessary database trips by specifying only updated relations when invoking `fresh()`? [The fresh method](https://laravel.com/docs/8.x/eloquent-collections#method-fresh) does support a list of relation names as an optional input.

thanks for reading!

**Comment by Gummibeer** [comment, marked as answer]
Hey,

thanks for your kind words and very descriptive issue! 🙏

The `fresh()` method is also in place for database default values - so dropping is not an option.

> Would it be feasible to prevent unnecessary database trips by specifying only updated relations when invoking fresh()? The fresh method does support a list of relation names as an optional input.

That's a very good idea/solution! We have a clear attribute/key pattern that identifies attributes to be logged as relation attributes. So it should be possible to match these, collect the relations, unify them and pass them to the `fresh()` method call.

Would you be willing/available to work on it? As you already have the database query count setup in place you could check if it has the wanted result.

**Reply by mac-s-g** [reply]
Yep, that makes sense.  
Sure, I can probably take a pass at it and post a PR sometime next week.
Thanks for reviewing!
